### ISSUE TITLE
tf.image.convert_image_dtype behaves incorrectly on Windows (but fine on MacOS)

### ISSUE THREAD

**Issue by petered** [tensorflow/tensorflow#60078]
<details><summary>Click to expand!</summary> 
 
 ### Issue Type

Bug

### Have you reproduced the bug with TF nightly?

No

### Source

source

### Tensorflow Version

2.9.1

### Custom Code

No

### OS Platform and Distribution

Windows

### Python version

3.10

### Current Behaviour?

Compare the result of the line 

```python
tf.image.convert_image_dtype(1., tf.uint32)
```
Between MacOS and Windows.  
* On Mac it returns `tf.Tensor(4294967295, shape=(), dtype=uint32)`
* On windows, `tf.Tensor(0, shape=(), dtype=uint32)`.

It seems that Mac has the correct result - you expect the result to saturate to the max uint32 value of `2**32-1==4294967295` when the float-input is >=1 - but it actually jumps back down to zero.  This caused a very sneaky bug in my code which caused the detection system to fail to detect in some cases.

### Standalone code to reproduce the issue

```python
assert int(tf.image.convert_image_dtype(1., tf.uint32))==2**32-1
```

### Relevant log output

_No response_</details>

**Comment by sushreebarsa** [comment]
@petered Thank you for reporting an issue!
Could you please provide a complete standalone code to replicate this issue?
Thank you!

**Comment by DEKHTIARJonathan** [comment]
Out of curiosity. Do you know if it changes anything with/without GPU? 

@sushreebarsa this should be enough for you to reproduce ...

```python
import tensorflow as tf
assert int(tf.image.convert_image_dtype(1., tf.uint32))==2**32-1
```

**Comment by petered** [comment]
Thanks for looking into it @sushreebarsa  and thanks @DEKHTIARJonathan  for responding.  Never tried with GPU.

**Comment by sushreebarsa** [comment]
@petered Could you please let us know the exact command you have used to install tensorflow here, it would help us to analyze the issue. Thank you!

**Comment by petered** [comment]
I believe it was just `pip install tensorflow` though I don't have the command-history available at the moment.  Are you not able to reproduce it?

**Comment by SuryanarayanaY** [comment]
@petered , @DEKHTIARJonathan 

The behaviour is due to the device the operation being executed.Please check the results on Colab(Linux environment) in the attached [gist](https://colab.research.google.com/gist/SuryanarayanaY/5171e789dd75342c0711d9f6405fc4dd/60078.ipynb) with CPU and GPU respectively. The issue is due to casting behaviour on CPU vs GPU with undefined/overflow values.

Please refer to the developer comment for similar issue [here](https://github.com/tensorflow/tensorflow/issues/58749#issuecomment-1467086661).

On MacOS by default the operation seems executing on GPU and on WIndows it is executing on CPU.Hence you are getting different results here.

Thankyou!

**Comment by petered** [comment]
Thanks @SuryanarayanaY for looking into that.  But it seems MacOS is getting the correct answer on CPU, not just GPU. (I have an M1 2020 Macbook air).  Running

```
import tensorflow as tf
print(tf.__version__)
with tf.device('CPU'):
  print('With CPU:')
  print(tf.image.convert_image_dtype(1., tf.uint32))
```
Prints:
```
2.9.1
With CPU:
tf.Tensor(4294967295, shape=(), dtype=uint32)
```

... Which is the "correct" answer.

**Comment by SuryanarayanaY** [comment]
This is due to undefined behaviour with cast overflow values. Please refer to this [comment](https://github.com/tensorflow/tensorflow/issues/58749#issuecomment-1467086661) for more context.

**Comment by petered** [comment]
Activity!

**Comment by SuryanarayanaY** [comment]
Hi @petered ,

The issue has been fixed in tf-nightly(2.14.0-dev20230503).Please refer to attached [gist](https://colab.research.google.com/gist/SuryanarayanaY/d168fa4aa643a64d269771ad75bfaa55/60078_r1.ipynb).

But there is one point to note here. Images that are represented using floating point values are expected to have values in the range [0,1).If you are using floats outside this range you need to set `saturate=True` in the API. 

Since floating point inputs to integer types may lead to over/underflow problems. Set saturate to True to avoid such problem in problematic conversions. If enabled, saturation will clip the output into the allowed range before performing a potentially dangerous cast. The same is documented now and you can refer the API Doc [source](https://www.tensorflow.org/api_docs/python/tf/image/convert_image_dtype) for more explanation.

Thank you!

**Comment by petered** [comment]
That's great, thank you for the fix.
